Verify if tunnels can be setup on Gateway node

sridhargaddam opened the pull request:
This PR validates if tunnels can be successfully established on the
Gateway node of local cluster. Validation is done by spawning a sniffer
pod on the local cluster Gateway node and a client pod on the non-Gateway
node of a remote cluster. The clientPod sends some UDP traffic to the
local Gateway node on the tunnelPort while tcpdump is executed in the
sniffer Pod. Finally, when both the Pods finish their execution, the
Pod output is compared with expected values and appropriate message
is displayed to the user.

Fixes issue: https://github.com/submariner-io/submariner-operator/issues/1147
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam committed:
Verify if tunnels can be setup on Gateway node

This PR validates if tunnels can be successfully established on the
Gateway node of local cluster. Validation is done by spawning a sniffer
pod on the local cluster Gateway node and a client pod on the non-Gateway
node of a remote cluster. The clientPod sends some UDP traffic to the
local Gateway node on the tunnelPort while tcpdump is executed in the
sniffer Pod. Finally, when both the Pods finish their execution, the
Pod output is compared with expected values and appropriate message
is displayed to the user.

Fixes issue: https://github.com/submariner-io/submariner-operator/issues/1147
Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

mangelajo commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+
+	defer sPod.DeletePod()
+	var gatewayPodIP string
+	if !localEndpoint.Spec.NATEnabled {
+		gatewayPodIP = localEndpoint.Spec.PrivateIP
+	} else {
+		gatewayPodIP = localEndpoint.Spec.PublicIP
+	}
```
This won't be accurate with the auto-detection mechanism. Perhaps we should try with both the Private & Public IPs and report which one worked.

sridhargaddam replied:
Done

mangelajo reviewed (changes requested):
Just one thing that needs handling.

mangelajo commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+)
+
+const (
+	ClientSourcePort = "9898"
```
Ok, I think this can be problematic. If we filter on source ports to detect that a packet arrives properly to the gateway, that can fail because the SRC port under some nat conditions can be modified.

As we were talking in slack, may be we can follow the next approach:

1) Try to find packets with src "ClientSourcePort", if that fails, push a WARNING about this
2) Try to find packets with the specific content (tcpdump matching on specific packet data), if this fails ERROR.

sounds ok? :)

sridhargaddam replied:
Yup. I'll update the PR.

tpantelis commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+
+var validateTunnelCmd = &cobra.Command{
+	Use:   "tunnel <localkubeconfig> <remotekubeconfig>",
+	Short: "Validate if firewall allows tunnels on the Gateway node.",
```
```suggestion
	Short: "Validate firewall access to Gateway node tunnels",
```

tpantelis commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+var validateTunnelCmd = &cobra.Command{
+	Use:   "tunnel <localkubeconfig> <remotekubeconfig>",
+	Short: "Validate if firewall allows tunnels on the Gateway node.",
+	Long:  "This command checks whether firewall configuration allows tunnel to be configured on the Gateway nodes.",
```
```suggestion
	Long:  "This command checks whether firewall configuration allows tunnels to be configured on the Gateway nodes.",
```

tpantelis commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+			return err
+		}
+		if same {
+			return fmt.Errorf("kubeconfig file <localkubeconfig> and <remotekubeconfig> need to have" +
+				" a unique content")
```
```suggestion
			return fmt.Errorf("the specified kubeconfig files are the same")
```

tpantelis commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+		}
+		return tunnelPort, nil
+	default:
+		message := fmt.Sprintf("Error parsing the tunnelPort for CableDriver %q",
+			submariner.Spec.CableDriver)
```
```suggestion
		message := fmt.Sprintf("Could not determine the tunnel port for cable driver %q",
			submariner.Spec.CableDriver)
```

tpantelis commented on pkg/subctl/cmd/validate_tunnel.go:
```diff
@@ -0,0 +1,186 @@
+	default:
+		message := fmt.Sprintf("Error parsing the tunnelPort for CableDriver %q",
+			submariner.Spec.CableDriver)
+		status.QueueWarningMessage(message)
```
```suggestion
		status.QueueFailureMessage(message)
```

sridhargaddam committed:
Modified the validation check to use nat-discovery ip

Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam commented:
@mangelajo @tpantelis addressed the review comments. Can you please take a re-look. Thanks.

sridhargaddam committed:
Ensure that validation works on AWS Cluster

On AWS Cluster the nodeNames do not match with the Hostname in
the localEndpoint. So this PR uses list of nodes from the cluster
and matches the node that has the same Hostname in localEndpoint.
We now spawn the snifferPod using the nodeName instead of Hostname.

Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

tpantelis committed:
Merge branch 'devel' into issue-1147
